dispatcher/dispatcher-impl/src/main/java/edu/tau/eng/neuroscience/mri/dispatcher/QueueManagerImpl.java

```diff
@@ -0,0 +1,213 @@
+                    Machine machine = machinesManager.getAvailableMachine();
+                    currTask.setMachine(machine);
+                    logger.debug("Allocated machine (machineId=" + machine.getId()
+                            + ") for task (taskId=" + currTask.getId() + ")");
+                    currTask.setStatus(TaskStatus.PROCESSING);
+                    logger.debug("Updating task's status in DB to 'processing' (taskId=" + currTask.getId() + ")");
+                    dbProxy.update(currTask);
+                    logger.debug("Sending task for execution (taskId=" + currTask.getId() + ")");
+                    executionProxy.execute(currTask);
+                }
+            } catch (InterruptedException e) {
+                logger.debug("Queue consumer thread shutting down after interrupt");
+            }
+        }
+    }
+
+    private class Producer extends Thread {
+
+        BlockingQueue<Task> queue;
+        DBProxy dbProxy;
+
+        public Producer(BlockingQueue<Task> queue, DBProxy dbProxy) {
+            this.queue = queue;
+            this.dbProxy = dbProxy;
+        }
+
+        @Override
+        public void run() {
+            try {
+                while (true) {
+                    synchronized (producerLock) {
+                        while (!producerMayWork) {
+                            producerLock.wait(MAX_PRODUCER_IDLE_TIME);
+                        }
+                    }
+                    List<Task> tasks = dbProxy.getNewTasks();
+                    synchronized(producerLock) {
+                        if (tasks == null || tasks.isEmpty()) {
+                            producerMayWork = false;
+                            continue;
+                        }
+                    }
+                    for (Task task: tasks) {
+                        Task currTask = shallowCopyTask(task);
+                        logger.debug("Adding new task to queue (taskId=" + currTask.getId() + ")");
+                        synchronized (consumerLock) {
+                            producerUpdatingDatabase = true;
+                        }
+                        boolean taskInsertedToQueue = queue.offer(currTask);
+                        if (!taskInsertedToQueue) {
+                            synchronized (consumerLock) {
+                                producerUpdatingDatabase = false;
+                                consumerLock.notify();
+                            }
+                            queue.put(currTask);
+                            synchronized (consumerLock) {
+                                producerUpdatingDatabase = true;
+                            }
+                        }
+                        task.setStatus(TaskStatus.PENDING);
+                        logger.debug("Updating task's status in DB to 'pending' (taskId=" + currTask.getId() + ")");
+                        dbProxy.update(currTask);
```

Comment:
If processing do not change to pending (specifically in this case since this is a new task)